### ISSUE TITLE
Decompose Storage

### ISSUE THREAD

**Issue by d-xo** [ethereum/hevm#231]
We currently use a nested array for storage in all smt queries. This is the most general model, but we can probably get some more performance by optimising this representation if the situation allows for it. I can see two cases:

1. There is only one address referenced in all storage reads & writes: We can model storage as a single (unnested) array.
2. There are multiple addresses, but they are all concrete: We can model storage as a collection of separate (unnested arrays).
3. All reads / writes refer to fully concrete slots and contractsaddresses: we can model storage as a collection of separate variables (i.e. no arrays needed).

2 & 3 can be mixed and matched (i.e. we can use arrays for contracts with writes / reads involving symbolic slots, and variables for contracts with fully concrete writes / reads).

**Comment by d-xo** [comment]
We may also be able to leverage some smt aided analysis to remove arrays from our queries, if we can e.g. prove that all reads / writes are disjoint within a given contracts storage

**Comment by msooseth** [comment]
Some notes about our discussion:
* We need to segregate arrays based on the 2nd 32B value of the ConcreteBuf into which we load the value before Keccak'ing it. This will almost always be a ConcreteBuf, unless it's a map of maps(?), or some hand-written assembly.
* For dynamic arrays, we can do a lookup for 0..255 static values and compute the offsets
* For the moment, we will resort back to unified array space when we see a constant value that we can't map back
* There is already a simplification pass to simplify storage reads that occur on top of a storage write.

Note to self: the simplification for the read-after-write may be a bit too strict. Below seems like may be a missed opportunity. Gotta check. Seemingly, we store `arg1` into both map positions 0 & 1, but only load from pos 0. Somehow, the pos1 is interfering(?).

```
        (SLoad
          slot:
            (Keccak
              (CopySlice
                srcOffset: 0x0
                dstOffset: 0x0
                size:      0x40
                src:
                  (WriteWord
                    idx:
                      0x0
                    val:
                      (Var "arg1")
                  )
                  (ConcreteBuf
                    Length: 64 (0x40) bytes
                    0000:   00 00 00 00  00 00 00 00  00 00 00 00  00 00 00 00   ................
                    0010:   00 00 00 00  00 00 00 00  00 00 00 00  00 00 00 00   ................
                    0020:   00 00 00 00  00 00 00 00  00 00 00 00  00 00 00 00   ................
                    0030:   00 00 00 00  00 00 00 00  00 00 00 00  00 00 00 00   ................
                  )
              )
              (ConcreteBuf "")
            )
          storage:
            (SStore
              slot:
                (Keccak
                  (CopySlice
                    srcOffset: 0x0
                    dstOffset: 0x0
                    size:      0x40
                    src:
                      (WriteWord
                        idx:
                          0x0
                        val:
                          (Var "arg1")
                      )
                      (ConcreteBuf
                        Length: 64 (0x40) bytes
                        0000:   00 00 00 00  00 00 00 00  00 00 00 00  00 00 00 00   ................
                        0010:   00 00 00 00  00 00 00 00  00 00 00 00  00 00 00 00   ................
                        0020:   00 00 00 00  00 00 00 00  00 00 00 00  00 00 00 00   ................
                        0030:   00 00 00 00  00 00 00 00  00 00 00 00  00 00 00 01   ................
                      )
                  )
                  (ConcreteBuf "")
                )
              val:
                (Var "arg2")
            )
            (SStore
              slot:
                (Keccak
                  (CopySlice
                    srcOffset: 0x0
                    dstOffset: 0x0
                    size:      0x40
                    src:
                      (WriteWord
                        idx:
                          0x0
                        val:
                          (Var "arg1")
                      )
                      (ConcreteBuf
                        Length: 64 (0x40) bytes
                        0000:   00 00 00 00  00 00 00 00  00 00 00 00  00 00 00 00   ................
                        0010:   00 00 00 00  00 00 00 00  00 00 00 00  00 00 00 00   ................
                        0020:   00 00 00 00  00 00 00 00  00 00 00 00  00 00 00 00   ................
                        0030:   00 00 00 00  00 00 00 00  00 00 00 00  00 00 00 00   ................
                      )
                  )
                  (ConcreteBuf "")
                )
              val:
                (Var "arg2")
            )
            (AbstractStore (SymAddr "entrypoint"))
        )
```

**Comment by msooseth** [comment]
## Notes to self

The above example has been resolved with #367 . The separate arrays are still not used, though. If we don't simplify as per #367 (which is not always possible), for the code:

```
        contract MyContract {
          mapping(uint => uint) items1;
          mapping(uint => uint) items2;
          function transfer(uint acct, uint val1, uint val2) public {
            uint beforeVal1 = items1[acct];
            uint beforeVal2 = items2[acct];
            unchecked {
              items1[acct] = val1+1;
              items2[acct] = val2+2;
              assert(items1[acct]+items2[acct] > beforeVal1 + beforeVal2);
            }
          }
        }
```


In the SMT2 we get:

```
# store4 = store into baseStore_symaddr_entrypoint at position: (keccak buf1) value: 1+arg2
(define-const store4 Storage (store baseStore_symaddr_entrypoint (keccak buf1) (bvadd (_ bv1 256) arg2)))

# store5 = store on top of store4 at position: (keccak buf3) value: 2+arg3
(define-const store5 Storage (store store4 (keccak buf3) (bvadd (_ bv2 256) arg3)))
```

This is too conservative. We could use 2 arrays here, one for `items1` and one for `items2`
